In [2]:
import numpy as np
print(f"numpy version: {np.__version__}")

numpy version: 1.22.3


## Part 1

The depositor deposited 50,000 u.e. in three different accounts in three different banks. 

Payments per annum:  
5% on the first account, 
7% on the second, 
and 6% on the third. 

It is known that in one year, the depositor received 2250 u.e. from the first and second banks 

and the sum of 1,400 u.e. from the first and third banks. 

How many u.e. did he put on each account initially?


So we got system of tree equesions:
x1 + x2 + x3 = 50000
0.5x1 + 0.7x2 + 0x3 = 2250
0.5x1 + 0x2 + 0.6x3 = 1400


In [39]:
f_matrix = np.array([[1, 1, 1], [0.05, 0.07, 0], [0.05, 0, 0.06]], dtype=np.float32)
f_matrix

array([[1.  , 1.  , 1.  ],
       [0.05, 0.07, 0.  ],
       [0.05, 0.  , 0.06]], dtype=float32)

In [40]:
f_array = np.array([50000, 2250, 1400], dtype=np.float32)
f_array

array([50000.,  2250.,  1400.], dtype=float32)

### Solutions

In [41]:
# Get inverse of a matrix
f_matrix_inv = np.linalg.inv(f_matrix)
f_matrix_inv

array([[ -1.8260869,  26.086954 ,  30.434782 ],
       [  1.3043478,  -4.347825 , -21.73913  ],
       [  1.5217391, -21.73913  ,  -8.695651 ]], dtype=float32)

In [45]:
# 1st check if determinant of an inverse matrix is not equal 0
det = round(np.linalg.det(f_matrix_inv))
det

-435

In [52]:
# get unswer via formula x = A^(-1)*B
x = f_matrix_inv.dot(f_array)
x.astype('int')

array([ 9999, 25000, 14999])

In [49]:
# one can also get solution via np.linalg.solve  (system of linear scalar equations)
x = np.linalg.solve(f_matrix, f_array)
x

array([10000., 25000., 15000.], dtype=float32)

The depositor placed: 
10,000 conditional u.e. on the first account, 
25,000 conditional u.e. on the second account, 
15,000 conditional u.e. on the third account. 


Checking the correctness of the answers: 

In [53]:
print(x.sum())

50000.0


In [56]:
print((x * f_matrix[1]).sum())

2250.0


In [61]:
print((x * f_matrix[2]).sum())

1399.9999


In [64]:
np.dot(f_matrix, x).all() == f_array.all()

True

# Part 2

iPhone 6, iPhone 11, iPhone 12 models are stored in the warehouse. 

A total of 1328 copies. 

iPhone 6 models are 120 fewer than iPhone 11 models, 

iPhone 6 models are 100 more than iPhone 12 models. 

How many models of each type are in stock?

Similar to the problem above:

x + y + z = 1328
-x + y + 0z = 120
x + 0y -z = 100

In [70]:
phone_matrix_a = np.matrix("1,1,1;-1,1,0;1,0,-1")
phone_matrix_a

matrix([[ 1,  1,  1],
        [-1,  1,  0],
        [ 1,  0, -1]])

In [71]:
phone_matrix_b = np.matrix("1328;120;100")
phone_matrix_b

matrix([[1328],
        [ 120],
        [ 100]])

In [72]:
# Get the numbers of each model 
phone_x = np.linalg.solve(phone_matrix_a, phone_matrix_b)
phone_x 

matrix([[436.],
        [556.],
        [336.]])

iPhone 6 - 436 pieces, 

iPhone 11 - 556 pieces, 

iPhone 12 models - 336 pieces are stored in the warehouse. 

Check the answer:

In [73]:
np.dot(phone_matrix_a, phone_x).all() == phone_matrix_b.all()

True

# Part 3

Write the canonical equation of the ellipsoid passing through three points A, B, C

Coordinates: 

A(sqr(3); 0; sqr(3))
B(sqr(6); 1/2; 0)
C(1; 1/sqr(3);1)

The canonical equation of an ellipsoid in a rectangular Cartesian coordinate system has the form:
x^2/a^2 + y^2/b^2 + z^2/c^2 = 1

Determining the parameters of a^2, b^2, c^2

Let's assume:

a^-2 = X1
b^-2 = X2
c^-2 = X3

We will get the system of equations:
3X1 + 0X2 + 3X3 = 1
6X1 + 1/4X2 + 0X3 = 1
X1 + 1/3X2 + X3 = 1

In [78]:
elipsoid_matrix = np.array([[3,0,3], [6,1/4,0],[1,1/3,1]], dtype=np.float32)
print(elipsoid_matrix)
elipsoid_array = np.array([1,1,1])

[[3.         0.         3.        ]
 [6.         0.25       0.        ]
 [1.         0.33333334 1.        ]]


In [79]:
elipsoid_x = np.linalg.solve(elipsoid_matrix, elipsoid_array)
elipsoid_x

array([0.08333334, 1.99999994, 0.25      ])

In [94]:
# get parameters
a2 = round(float(1/elipsoid_x[0]), 1)
b2 = round(float(1/elipsoid_x[1]), 1)
c2 = round(float(1/elipsoid_x[2]), 1)
print(f'Parameters: a2={a2}, b2={b2}, c2={c2}')

Parameters: a2=12.0, b2=0.5, c2=4.0


Check the answers:

In [96]:
# Check Coordinate A
x = 3**0.5
y=0
z=3**0.5
round(x**2/a2 + y**2/b2 + z**2/c2, 2) == 1

True

In [95]:
# Check Coordinate B
x = 6**0.5
y=1/2
z=0
round(x**2/a2 + y**2/b2 + z**2/c2, 2) == 1

True

In [97]:
# Check Coordinate C
x = 1
y=1/(3**0.5)
z=1
round(x**2/a2 + y**2/b2 + z**2/c2, 2) == 1

True

The canonical equation of an ellipsoid:
x^2/12 + 2y^2 + z^2/4 = 1

# Part 4

Write the parabola equation that passes through the three points (1,12), (3,54), (-1,2). 

Find the coefficients a,b,c of the equation:

y=a⋅x^2+b⋅x+c


Let us add the y and x values from the points to the equation:

a + b + c = 12
9a + 3b + c = 54
a - b + c = 2

In [99]:
parabola_matrix = np.array([[1,1,1],[9,3,1],[1,-1,1]], np.float32)
parabola_matrix

array([[ 1.,  1.,  1.],
       [ 9.,  3.,  1.],
       [ 1., -1.,  1.]], dtype=float32)

In [100]:
parabola_array = np.array([12,54,2], np.float32)
parabola_array

array([12., 54.,  2.], dtype=float32)

In [101]:
parabola_x = np.linalg.solve(parabola_matrix, parabola_array)
parabola_x

array([4., 5., 3.], dtype=float32)

In [105]:
print(f'Coefficients a={int(parabola_x[0])}, b={int(parabola_x[1])}, c={int(parabola_x[2])}')

Coefficients a=4, b=5, c=3


The parabola equation: y = 4x^2 + 5x + 3

In [107]:
#Check the answer:
np.dot(parabola_matrix, parabola_x).all() == parabola_array.all()

True

# Part 5

Create polynomial function `get_polynom`, which based on a set of points (x1,y2), (x2,y2),...,(xn+1,yn+1), will give
the coefficient `ci`. 

Thus, the function accepts a list of tuples with coordinates and returns a set of coefficients c0, c1,...,cn

The polynomial function: P(x) = c0 + c1x^1 + c2x^2 + c3x^3 + ... + cnx^n
P(x) = y, degree n-1 polynomials have the right length for n points.
For n given points we will get n equations, like:
c0 + c1x1 c2x1^2 c3x1^3 ... cnx1^n = y1
.
.
.

The unswer: 'Av = b'
from the coordinats take x value and build matrix A using polynomial function:

1 x1 x1^2 x1^3 ... x1^n
1 x2 x2^2 x2^3 ... x2^n
1 x3 x3^2 x3^3 ... x3^n
.
.
.
1 xn xn^2 xn^3 ... xn^n

multiply this matrix by vector of coeficient:
v = (c0, c1, c2, c3, ..., cn)

and get the vector of knovns b:
b = (y1, y2, y3, ..., yn)

In [168]:
def get_polynom(coords):
    n = len(coords)
    x_matrix_list = [] 
    y_known_list = []
    
    if n > 1:
        for x,y in coords:
            # get the vector of knowns 
            y_known_list.append(y)
            # build the matrix from n equations
            x_matrix_row_list = []
            for i in range(n):
                x_matrix_row_list.append(pow(x,i))
            x_matrix_list.append(x_matrix_row_list)
            
        A = np.array(x_matrix_list)   
        b = np.array(y_known_list) 
        print(f'Matrix with knowns: \n A={A}')
        print(f'Vector of y: b={b}')
        v = np.linalg.solve(A, b)
        print(f'Vector of coefficients v: v={v}')
        n = 0
        for c in v:
            print(f'C{n} = {round(c,2)}')
            n += 1 
        
        print(f"Check the answer: Av = b is {np.dot(A, v).all() == b.all()}")
        return v
    else:
        return 'The polynomial degree cannot be less than 1'
        

In [169]:
coord_list = [(10,4), (0,10), (5, 121), (6, 125)]

coord = get_polynom(coord_list)

Matrix with knowns: 
 A=[[   1   10  100 1000]
 [   1    0    0    0]
 [   1    5   25  125]
 [   1    6   36  216]]
Vector of y: b=[  4  10 121 125]
Vector of coefficients v: v=[10.         25.91666667  1.165      -0.38166667]
C0 = 10.0
C1 = 25.92
C2 = 1.16
C3 = -0.38
Check the answer: Av = b is True
